### ISSUE TITLE
Aes/Pclmulqdq IsSupported should depend on COMPlus_EnableSSE3_4

### ISSUE THREAD

**Issue by fiigii** [dotnet/coreclr#20468]
`AES` and `PCLMULQDQ` instructions use the three-byte opcode (`0x0F38` or `0x0F3A`) encoding with the mandatory prefix `0x66`, which is an SSSE3 feature. Meanwhile, all the processors (Intel and AMD both) that support `AES` and `PCLMULQDQ` have SSSE3 ISA.

Related to https://github.com/dotnet/coreclr/issues/20461

@CarolEidt @tannergooding

**Comment by fiigii** [comment]
I will submit a PR to fix this.

**Comment by tannergooding** [comment]
This does not match up with what the architecture guides specify:
![image](https://user-images.githubusercontent.com/10487869/47120973-b198bc80-d225-11e8-99b4-4dd5efcef3bb.png)

**Comment by fiigii** [comment]
Yes, but the encoding (`0x660F38/A`) is only used by SSSE3 and above ISAs (not SSE4.1, I have corrected the issue description). And RuyJIT also treats the encoding as a `SSE3_4` feature.

**Comment by tannergooding** [comment]
that sounds like something we should likely fixup in the JIT.

**Comment by fiigii** [comment]
Or we can just change `Aes/Pclmulqdq.IsSupported` depends on the jit knob `COMPlus_EnableSSSE3` and `COMPlus_EnableSSE3_4` but still drived from `Sse2` in API.

**Comment by tannergooding** [comment]
At the same time, I don't think it is worthwhile to deviate from the spec and what most systems will be doing to check if a given feature is supported.

The spec says to check for SSE2 and for the Individual ISA flag, and that is what the majority of systems wanting to check for this feature support will be doing. I don't think we should be any different.

For example, you can check what flags that clang/gcc expect to be passed to use certain intrinsics, and they will not let you use SSE3 intrinsics just because `-mpclmul` is specified: https://godbolt.org/z/9U_1cL

**Comment by tannergooding** [comment]
> Or we can just change Aes/Pclmulqdq.IsSupported depends on the jit knob COMPlus_EnableSSSE3 and COMPlus_EnableSSE3_4 but still drived from Sse2 in API.

This seems like a reasonable workaround for the issue, but I would think we want these flags more generally cleaned up or fixed to be clear on what they control.

**Comment by tannergooding** [comment]
> Currently, RuyJIT specially treats 0x660F38/A encoding whenIs4ByteSSE4Instruction or Is4ByteSSE4OrAVXInstruction returns true.

Right, these methods (in general) are incorrectly named as the 4-byte encoding is not limited to SSE4 (it also exists for certain SSSE3 instructions) and most of the `Is*AVXInstruction` checks are really "is this a VEX encoded instruction".

This should just be cleanup in the emitter to adjust this information. Ideally we could also adjust the COMPlus flags to be more clear (and we've discussed that a bit on another issue), but that is more likely to be a "breaking change" (at least for some code, since it impacts an existing "retail" flag).

**Comment by fiigii** [comment]
Just to clarify again, there would be two solutions
1. let RyuJIT emitter work for `0x660F38/A` encoding even if  the current `Is4ByteSSE4Instruction` returns `false` (because of `COMPlus_EnableSSE3_4=0`)
2. let `Aes/Pclmulqdq.IsSupported` return `false` when `COMPlus_EnableSSE3_4=0`, which is https://github.com/dotnet/coreclr/pull/20469

**Comment by tannergooding** [comment]
I think 1 should be:
* Update the emitter so that `COMPlus_EnableSSE3_4` only impacts SSE3 and higher. Any ISAs which are not spec'd as being dependent on SSE3 would "just work", regardless of their encoding

We should additionaly
* Update the emitter so that `COMPlus_EnableAVX` only impacts AVX and higher. Any ISAs which are not spec'd as being dependent on AVX would "just work", regardless of their encoding.

**Comment by CarolEidt** [comment]
I'm happy with this for now.
I agree that it would be good to separately control the encoding. Not so much because I think we need to worry about supporting a hypothetical case where `AES` and `PCLMULQDQ` are supported independent of `SSSE3`, but rather because it would be a cleaner solution.
I'm not sure I understand @tannergooding 's second point - that is, are there instructions that use AVX encodings that aren't in a dependent ISA?

**Comment by CarolEidt** [comment]
Oops - meant the "I'm happy with this for now" to be on #20469

**Comment by tannergooding** [comment]
@CarolEidt, yes. BMI1/2 depend on the VEX encoding but do not depend on AVX.

I dont believe we will ever encounter a processor that supports one without the other, but it is possible to disable the YMM state support at the OS level, in which case BMI1/BMI2 are useable but AVX is not (hence why these ISAs do not inherit from AVX).
